### PR TITLE
Namespace selector

### LINKED ISSUE #202: Add namespace selection UI under

### PR DESCRIPTION
**Pull request by fabiovincenzi**
fixes #202

### CODE REVIEW

File: pkg/api/admin/dao/repositories/namespace.go [hunk 1]
```diff
@@ -51,3 +53,12 @@ func (r NamespaceRepository) GetByCode(ctx context.Context, code string) (*model
 	}
 	return &namespace, nil
 }
+
+// ListNamespaces returns a list of all namespaces.
+func (r NamespaceRepository) ListNamespaces(ctx context.Context) ([]models.Namespace, error) {
```
Comment (dave-gantenbein) [inline pkg/api/admin/dao/repositories/namespace.go:58]: What do we think of the stuttering implied by this naming convention, ie "namespaces" appearing twice: namespaceRepository.ListNamespaces. Maybe #List would be clear enough and shorter?

File: pkg/api/admin/service/namespace/service.go [hunk 2]
```diff
@@ -22,5 +22,9 @@ func NewService(
 }
 
 func (s Service) ListNamespaces(ctx context.Context) ([]models.Namespace, error) {
```
Comment (dsuhinin) [inline pkg/api/admin/service/namespace/service.go:24]: it is pretty simple now but we can create a `unit` tests here. what do you think?

File: pkg/ui/chooser/routes.go [hunk 3]
```diff
@@ -8,21 +8,45 @@ import (
...
+
+func (r Router) AddRoutes(fr fiber.Router) {
 	sub, _ := fs.Sub(content, "embed")
 
-	r.Use("/static/chooser/", etag.New(), filesystem.New(filesystem.Config{
+	fr.Use("/static/chooser/", etag.New(), filesystem.New(filesystem.Config{
 		Root: http.FS(sub),
 	}))
+	// engine and app for template rendering
+	engine := html.NewFileSystem(http.FS(sub), ".html")
```
Comment (dsuhinin) [inline pkg/ui/chooser/routes.go:39]: can we put initialisation right into:
```
app := fiber.New(fiber.Config{
 		Views: html.NewFileSystem(http.FS(sub), ".html"),
 	})
```

File: pkg/api/admin/dao/repositories/namespace.go [hunk 4]
```diff
@@ -16,6 +16,8 @@ type NamespaceRepositoryProvider interface {
 	Create(ctx context.Context, namespace *models.Namespace) error
 	// GetByCode returns namespace by its Code.
 	GetByCode(ctx context.Context, code string) (*models.Namespace, error)
+	// ListNamespaces returns all the namespaces
+	ListNamespaces(ctx context.Context) ([]models.Namespace, error)
```
Comment (dsuhinin) [inline pkg/api/admin/dao/repositories/namespace.go:20]: yep, we can simplify it to be just `List`.

**Review by dsuhinin — CHANGES_REQUESTED** [review]

File: pkg/ui/chooser/routes.go [hunk 5]
```diff
@@ -8,21 +8,45 @@ import (
...
+}
+
+// NewRouter creates new instance of `mlflow` router.
+func NewRouter(controller *controller.Controller) *Router {
+	return &Router{
+		controller: controller,
+	}
+}
+
+func (r Router) AddRoutes(fr fiber.Router) {
```
[the remark this answers is not in the document]
Comment (fabiovincenzi) [reply pkg/ui/chooser/routes.go:32]: Yes!

File: pkg/api/admin/service/namespace/service.go (hunk 2 above)
Comment (fabiovincenzi) [reply pkg/api/admin/service/namespace/service.go:24]: yep, I'll do it

File: pkg/api/admin/dao/repositories/namespace.go (hunk 1 above)
Comment (fabiovincenzi) [reply pkg/api/admin/dao/repositories/namespace.go:58]: Yep!

**Review by dsuhinin — APPROVED** [review]
Looks good for me. but will be nice to have other eyes on that. @jgiannuzzi

**Review by dave-gantenbein — APPROVED** [review]

File: pkg/ui/chooser/routes.go [hunk 6]
```diff
@@ -8,23 +8,52 @@ import (
...
+	// specific routes
+	app.Get("/", r.controller.GetNamespaces)
+
+	// default route
+	app.Use("/", etag.New(), filesystem.New(filesystem.Config{
 		Root: http.FS(
 			common.NewOnlyRootFS(sub, "index.html"),
 		),
 	}))
+	return nil
```
Comment (jgiannuzzi) [inline pkg/ui/chooser/routes.go:58]: there is an overlap between these various routes and the last one does not get used

File: pkg/api/admin/service/namespace/service.go [hunk 7]
```diff
@@ -21,6 +21,11 @@ func NewService(
 	}
 }
 
+// ListNamespaces returns all namespaces.
 func (s Service) ListNamespaces(ctx context.Context) ([]models.Namespace, error) {
-	return nil, nil
+	namespaces, err := s.namespaceRepository.List(ctx)
```
[the remark this answers is not in the document]
Comment (dsuhinin) [reply pkg/api/admin/service/namespace/service.go:26]: this is probably something old. maybe It seems like I left it after my tests. yep, please remove. fresh branches have no such change.

File: pkg/ui/chooser/routes.go [hunk 8]
```diff
@@ -8,21 +8,45 @@ import (
...
+
+// NewRouter creates new instance of `mlflow` router.
+func NewRouter(controller *controller.Controller) *Router {
+	return &Router{
+		controller: controller,
+	}
+}
+
+func (r Router) AddRoutes(fr fiber.Router) {
 	sub, _ := fs.Sub(content, "embed")
```
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply pkg/ui/chooser/routes.go]: as I commented below, this cannot give an error as it just wouldn't build if there was no "embed" top directory

[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply pkg/ui/chooser/routes.go]: so I think we should keep skipping the error so we don't complicate the code further

[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply pkg/ui/chooser/routes.go]: sorry for the flip-flop @fabiovincenzi!

File: pkg/ui/chooser/embed/index.html [hunk 9]
```diff
@@ -56,12 +63,34 @@
...
+            var namespaceText = document.getElementById('selected-namespace')
+            if (namespaceIndex !== -1) {
+                namespaceText.innerText = "Selected Namespace: " + currentUrl.substring(namespaceIndex + 4).slice(0, -1);
+            } else {
+                namespaceText.innerText = "Selected Namespace: default"
+            }
+        }
+    
+        window.onload = setCurrentNamespace;
+    </script>
```
[the remark this answers is not in the document]
Comment (fabiovincenzi) [reply pkg/ui/chooser/embed/index.html:93]: How do we know the selected namespace from the backend?

File: pkg/api/admin/dao/repositories/mock_namespace_repository_provider.go
Comment (jgiannuzzi) [inline pkg/api/admin/dao/repositories/mock_namespace_repository_provider.go:1]: please add this provider to `.mockery.yaml`

File: pkg/ui/chooser/embed/index.html (hunk 9 above)
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply pkg/ui/chooser/embed/index.html:93]: by using `namespace.GetNamespaceFromContext` like we do in every controller

**Review by jgiannuzzi — APPROVED** [review]

**Comment by jgiannuzzi** [comment]
future improvement ideas:
- show the active namespace in the top right corner
- when this gets clicked, show the list of namespaces
- (maybe) when there is only one namespace, don't display any namespace-related stuff

**Pull request merged**